### PR TITLE
Impute tips

### LINKED ISSUE #215: Impute tips
### Why this matters  
* **Policy coverage:** Proposals such as “No Tax on Tips” require a clean `tip_income` field distinct from regular wages.  
* **Distributional accuracy:** Tipped workers skew lower-income; omitting tips biases poverty, EITC/CTC, and payroll-tax results.  
* **Benchmark alignment:** IRS publishes line-item totals for unreported and allocated tips, so our microdata should reconcile to those figures.

---

### IRS control totals (calibration targets)

| Field | Tax year | Amount (billions) | Meaning | Source |
|-------|----------|-------------------|---------|--------|
| **Form 1040 Line 1c – “Tip income not reported on line 1a”** | 2022 | **145.357** | Tips that filers self-report via Form 4137 (i.e., **unreported** to employers) | Publication 4801: *Individual Income-Tax Returns Line-Item Estimates, 2022* <https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-return-line-item-estimates> |
| **Form W-2 Box 8 – Allocated tips** | 2020 | **0.048** | Tips employers add when reported tips < 8 % of gross receipts | Publication 5385: *Individual Information Returns Line-Item Estimates, 2020* <https://www.irs.gov/statistics/soi-tax-stats-individual-information-return-line-item-estimates> |

**Calibration rule**  
`sum_cps(tip_income)` must equal **$145.405 billion** (= Line 1c + Box 8) for the target tax year after adjusting for non-filers via our usual filing-probability weights.  
*(W-2 Box 7 “social-security tips,” already embedded in wages, stays untouched.)*

---

### Data and modeling plan  

1. **Training data**  
   * Public-use SIPP panels 2017-2024 (national).  

2. **Model** – single stage  
   * Use **`microimpute`** QRF to predict **annual dollar tips** (`tip_income_sipp`) directly.  

4. **Calibration loop**  
   * Target the IRS total

### References  

1. IRS SOI. *Individual Income-Tax Returns Line-Item Estimates, 2022.* <https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-return-line-item-estimates>  
2. IRS SOI. *Individual Information Returns Line-Item Estimates, 2020.* <https://www.irs.gov/statistics/soi-tax-stats-individual-information-return-line-item-estimates>  
3. Yale Budget Lab (2024). *No Tax on Tips – Technical Appendix.* <https://budgetlab.yale.edu/research/no-tax-tips-budgetary-distributional-and-tax-avoidance-considerations>  
4. U.S. Census Bureau (2021). *SIPP User Note: Changes to Imputed Earnings.* <https://www.census.gov/programs-surveys/sipp/tech-documentation/user-notes/2020-usernotes/chngs-imputd-earngs.html>

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
Fixes #215

### CODE REVIEW

**Comment by nikhilwoodruff** [comment]
This is person-level: 
image

**Comment by MaxGhenis** [comment]
Do others use the SIPP for overtime? Yale does it from the CPS ORG: https://github.com/PolicyEngine/policyengine-us-data/issues/217

File: policyengine_us_data/datasets/sipp/sipp.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            repo_id="PolicyEngine/policyengine-us-data",
+            filename="pu2023_slim.csv",
+            repo_type="model",
+            local_dir=STORAGE_FOLDER,
+        )
+        df = pd.read_csv(
+            STORAGE_FOLDER / "pu2023_slim.csv",
+        )
+
+    df["tip_income"] = (
```
Comment (MaxGhenis) [inline policyengine_us_data/datasets/sipp/sipp.py:88]: Is this right?
```suggestion
    # Sum tip columns (AJB*_TXAMT + TJB*_TXAMT) across all jobs.
    df["tip_income"] = (
```

**Review by MaxGhenis — APPROVED** [review]
Some minor suggestions and questions, good to merge though

File: policyengine_us_data/datasets/sipp/sipp.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            sipp.index,
+            size=100_000,
+            replace=True,
+            p=sipp.household_weight / sipp.household_weight.sum(),
+        )
+    ]
+
+    model = QRF()
+
+    model = model.fit(
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply policyengine_us_data/datasets/sipp/sipp.py:111]: Yep filed! https://github.com/PolicyEngine/microimpute/issues/49

File: policyengine_us_data/datasets/sipp/sipp.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    df["employment_income"] = df.TPTOTINC * 12
+    df["is_under_18"] = (df.TAGE < 18) & (df.MONTHCODE == 12)
+    df["is_under_6"] = (df.TAGE < 6) & (df.MONTHCODE == 12)
+    df["count_under_18"] = (
+        df.groupby("SSUID")["is_under_18"].sum().loc[df.SSUID.values].values
+    )
+    df["count_under_6"] = (
+        df.groupby("SSUID")["is_under_6"].sum().loc[df.SSUID.values].values
+    )
+    df["household_weight"] = df.WPFINWGT / 12
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply policyengine_us_data/datasets/sipp/sipp.py]: Removed actually (makes no difference and I think it was erroneous).

File: policyengine_us_data/datasets/sipp/sipp.py (hunk 1 above)
Comment (nikhilwoodruff) [reply policyengine_us_data/datasets/sipp/sipp.py:69]: Added, yes.

File: policyengine_us_data/datasets/sipp/sipp.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            "TPTOTINC",
+        ]
+
+        for col in cols:
+            if "JB1" in col:
+                for i in range(2, 8):
+                    cols.append(col.replace("JB1", f"JB{i}"))
+
+        df = pd.read_csv(
+            "/Users/nikhilwoodruff/Downloads/pu2023.csv",
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply policyengine_us_data/datasets/sipp/sipp.py]: Fixed.

File: policyengine_us_data/datasets/sipp/sipp.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            "EDEPCLM",
+            "EMS",
+            "EFSTATUS",
+            "TJB1_TXAMT",
+            "TJB1_MSUM",
+            "TJB1_OCC",
+            "TJB1_IND",
+            "AJB1_TXAMT",
+            "EJB1_TYPPAY3",
+            "TJB2_TXAMT",
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply policyengine_us_data/datasets/sipp/sipp.py]: Yes, removed.

File: policyengine_us_data/storage/download_public_prerequisites.py [hunk 6]
```diff
@@ -1,6 +1,14 @@
 from policyengine_us_data.utils.github import download
 from pathlib import Path
 from policyengine_us_data.storage import STORAGE_FOLDER
+from huggingface_hub import hf_hub_download
+
+hf_hub_download(
```
Comment (nikhilwoodruff) [inline policyengine_us_data/storage/download_public_prerequisites.py]: Removing this.

File: policyengine_us_data/utils/loss.py [hunk 7]
```diff
@@ -256,6 +256,8 @@ def build_loss_matrix(dataset: type, time_period):
         # Rough estimate, not CPS derived
         "real_estate_taxes": 500e9,  # Rough estimate between 350bn and 600bn total property tax collections
         "rent": 735e9,  # ACS total uprated by CPI
+        "tip_income": 38e9
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply policyengine_us_data/utils/loss.py]: OK fixed!

File: policyengine_us_data/utils/loss.py [hunk 8]
```diff
@@ -256,6 +256,8 @@ def build_loss_matrix(dataset: type, time_period):
...
+        * 1.35,  # Table 5A from https://www.irs.gov/statistics/soi-tax-stats-individual-information-return-form-w2-statistics, uprated by 35% employment income growth 2018-2024.
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply policyengine_us_data/utils/loss.py]: Fixed.

**Review by MaxGhenis — APPROVED** [review]

**Review by PavelMakarchuk — APPROVED** [review]
